Content.Shared/Verbs/SharedVerbSystem.cs

```diff
@@ -1,62 +1,139 @@
 
 namespace Content.Shared.Verbs
 {
     public class SharedVerbSystem : EntitySystem
     {
+        [Dependency] private readonly IEntityLookup _entityLookup = default!;
+
         /// <summary>
-        ///     Get all of the entities relevant for the contextmenu
+        ///     Get all of the entities in an area for displaying on the context menu.
         /// </summary>
-        /// <param name="player"></param>
-        /// <param name="targetPos"></param>
-        /// <param name="contextEntities"></param>
-        /// <param name="buffer">Whether we should slightly extend out the ignored range for the ray predicated</param>
-        /// <returns></returns>
-        public bool TryGetContextEntities(IEntity player, MapCoordinates targetPos, [NotNullWhen(true)] out List<IEntity>? contextEntities, bool buffer = false)
+        /// <param name="buffer">Whether we should slightly extend the entity search area.</param>
+        public bool TryGetContextEntities(IEntity player, MapCoordinates targetPos,
+            [NotNullWhen(true)] out List<IEntity>? contextEntities, bool buffer = false, bool ignoreVisibility = false)
         {
             contextEntities = null;
-            var length = buffer ? 1.0f: 0.5f;
 
-            var entities = IoCManager.Resolve<IEntityLookup>().
-                GetEntitiesIntersecting(targetPos.MapId, Box2.CenteredAround(targetPos.Position, (length, length))).ToList();
+            // Check if we have LOS to the clicked-location.
+            if (!ignoreVisibility && !player.InRangeUnOccluded(targetPos, range: ExamineSystemShared.ExamineRange))
+                return false;
+
+            // Get entities
+            var length = buffer ? 1.0f : 0.5f;
+            var entities = _entityLookup.GetEntitiesIntersecting(targetPos.MapId,
+                Box2.CenteredAround(targetPos.Position, (length, length))).ToList();
 
             if (entities.Count == 0)
+                return false;
+
+            if (ignoreVisibility)
             {
+                contextEntities = entities;
+                return true;
+            }
+
+            // perform visibility checks
+            foreach (var entity in entities.ToList())
+            {
+                if (entity.HasTag("HideContextMenu") ||
+                    !player.InRangeUnOccluded(entity, range: ExamineSystemShared.ExamineRange))
+                {
+                    entities.Remove(entity);
+                }
+            }
+
+            if (entities.Count == 0)
                 return false;
+
+            contextEntities = entities;
+            return true;
+        }
+
+        /// <summary>
+        ///     Raises a number of events in order to get all verbs of the given type(s)
+        /// </summary>
+        public Dictionary<VerbType, List<Verb>> GetVerbs(IEntity target, IEntity user, VerbType verbTypes)
+        {
+            Dictionary<VerbType, List<Verb>> verbs = new();
+
+            if (verbTypes.HasFlag(VerbType.Activation))
```

Comment:
don't use HasFlag, just do `verbTypes & VerbType.Activation` or whatever bitwise operation you need

https://stackoverflow.com/questions/7368652/what-is-it-that-makes-enum-hasflag-so-slow

Reply:
Changed. Also fixed a Activation vs Interaction typo